test/clj/rems/application/test_process_managers.clj

```diff
@@ -0,0 +1,190 @@
+                  (is (= [{:attachment/id attachment-id1 :attachment/filename "attachment1.txt" :attachment/type "text/plain"}]
+                         (:application/attachments (applications/get-application-internal app-id))
+                         (attachments/get-attachments-for-application app-id))
+                      "attachment1 was saved"))
+
+                (testing "upload attachment2"
+                  (let [attachment-id2 (upload-request app-id "alice" "attachment2.txt")]
+
+                    (testing "use attachment2 in a field"
+                      (is (= {:success true}
+                             (-> (request :post (str "/api/applications/save-draft" ))
+                                 (authenticate "42" "alice")
+                                 (json-body {:application-id app-id
+                                             :field-values [{:form form-id :field "attachment1" :value (str attachment-id1)}
+                                                            {:form form-id :field "attachment2" :value (str attachment-id2)}]})
+                                 handler
+                                 read-ok-body)))
+
+                      (is (= [{:attachment/id attachment-id1 :attachment/filename "attachment1.txt" :attachment/type "text/plain"}
+                              {:attachment/id attachment-id2 :attachment/filename "attachment2.txt" :attachment/type "text/plain"}]
+                             (:application/attachments (applications/get-application-internal app-id))
+                             (attachments/get-attachments-for-application app-id))
+                          "attachment1 and attachment2 are saved"))
+
+                    (testing "upload attachment3"
+                      (let [_attachment-id3 (upload-request app-id "alice" "attachment3.txt")]
+
+                        (testing "send application"
+                          ;; NB: don't save again, so attachment3 shouldn't be in use
+
+                          (is (= {:success true}
+                                 (-> (request :post "/api/applications/submit")
+                                     (authenticate "42" "alice")
+                                     (json-body {:application-id app-id})
+                                     handler
+                                     read-ok-body)))
+
+                          ;; NB: attachment 3 should have been cleaned
+
+                          (is (= [{:attachment/id attachment-id1 :attachment/filename "attachment1.txt" :attachment/type "text/plain"}
+                                  {:attachment/id attachment-id2 :attachment/filename "attachment2.txt" :attachment/type "text/plain"}]
+                                 (:application/attachments (applications/get-application-internal app-id))
+                                 (attachments/get-attachments-for-application app-id))
+                              "attachment1 and attachment2 are saved, but not attachment3"))))
```

Comment:
should we also verify that the attachment has been removed from attachments table?

Reply:
Now that I look at it, that's exactly what `get-attachments-for-application` does, isn't it?